extensions/configoperations/permutations/permutations.go

```diff
@@ -0,0 +1,107 @@
+package permutations
+
+import (
+	"encoding/json"
+
+	"github.com/rancher/shepherd/extensions/configoperations"
+)
+
+// Relationship structs are used to create an association between a parent value and either a set of permutations or the value of a different key
+type Relationship struct {
+	ParentValue       any           `json:"parentValue" yaml:"parentValue"`
+	ChildKeyPath      []string      `json:"childKeyPath" yaml:"childkeyPath"`
+	ChildKeyPathValue any           `json:"childKeyPathValue" yaml:"childkeyPathValue"`
+	ChildPermutations []Permutation `json:"childPermutations" yaml:"childPermutations"`
+}
+
+// Permutation structs are used to describe a single permutation
+type Permutation struct {
+	KeyPath                   []string       `json:"keyPath" yaml:"keyPath"`
+	KeyPathValues             []any          `json:"keyPathValue" yaml:"keyPath"`
+	KeyPathValueRelationships []Relationship `json:"keyPathValueRelationships" yaml:"KeyPathValueRelationships"`
+}
+
+// CreateRelationship is a constructor for the relationship struct
+func CreateRelationship(parentValue any, childKeyPath []string, childKeyPathValue any, childPermutations []Permutation) Relationship {
+	return Relationship{
+		ParentValue:       parentValue,
+		ChildKeyPath:      childKeyPath,
+		ChildKeyPathValue: childKeyPathValue,
+		ChildPermutations: childPermutations,
+	}
+}
+
+// CreatePermutation is a constructor for the permutation struct
+func CreatePermutation(keyPath []string, keyPathValues []any, keyPathValueRelationships []Relationship) Permutation {
+	return Permutation{
+		KeyPath:                   keyPath,
+		KeyPathValues:             keyPathValues,
+		KeyPathValueRelationships: keyPathValueRelationships,
+	}
+}
+
+// Permute iterates over a list of permutation structs and permutes the base config with each of the permutations
+func Permute(permutations []Permutation, baseConfig map[string]any) ([]map[string]any, string, error) {
```

Comment:
what is the returned string used for? I only see it set to an empty string in all the returns

Reply:
Your right 😓 I forgot to update that. It is going to return a list of permutation names.

Reply:
After considering further I have decided to remove this functionality from the Permute function. If we want to build the permutation name there are other ways to do so that are more effective.